### PR TITLE
Changes for dynamic workflow execution

### PR DESCRIPTION
**Pull request by savan77**
These updates may get changed based on front-end. 

### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2020 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: cvat/apps/onepanelio/views.py [hunk 1]
```diff
@@ -4,28 +4,84 @@
...
-    configuration.api_key['authorization'] = auth_token
-    # Uncomment below to setup prefix (e.g. Bearer) for API key, if needed
-    configuration.api_key_prefix['authorization'] = 'Bearer'
 
-    # Defining host is optional and default to http://localhost:8888
-    configuration.host = os.getenv('ONEPANEL_API_URL')
 
+def onepanel_authorize():
+    # auth_token = AuthToken.get_auth_token(request)
+    auth_token = os.getenv('ONEPANEL_AUTHORIZATION')
```
Comment (savan77) [inline cvat/apps/onepanelio/views.py]: This is for development purpose. Line 36 will be used in production.

File: cvat-ui/src/onepanel/createAnnotationModal/createAnnotaion.component.tsx [hunk 2]
```diff
@@ -16,14 +16,10 @@ import {
     Spin
```
Comment (rushtehrani) [inline cvat-ui/src/onepanel/createAnnotationModal/createAnnotaion.component.tsx]: File name is misspelled, should be `createAnnotation`

Comment (savan77) [reply cvat-ui/src/onepanel/createAnnotationModal/createAnnotaion.component.tsx]: Noted.

File: cvat/apps/onepanelio/views.py [hunk 3]
```diff
@@ -39,3 +95,248 @@ def get_workflow_templates(request):
...
+
+        if cloud_provider == "s3":
+            
+            import boto3
+            from botocore.exceptions import ClientError
+
+            #check if datasets folder exists on aws bucket
+            s3_client = boto3.client('s3')
+
+            try:
```
Comment (savan77) [inline cvat/apps/onepanelio/views.py:212]: Just realized, this try/except block may not be needed.

File: cvat-ui/src/onepanel/createAnnotationModal/createAnnotation.container.tsx [hunk 4]
```diff
@@ -6,17 +6,14 @@ import React from 'react';
 import { connect } from 'react-redux';
 
 // import ModelRunnerModalComponent from 'components/model-runner-modal/model-runner-modal';
-import ModelNewAnnotationModalComponent from 'components/model-new-annotation-modal/model-new-annotation-modal';
+import ModelNewAnnotationModalComponent from './createAnnotaion.component';
```
Comment (aleksandrmelnikov) [inline cvat-ui/src/onepanel/createAnnotationModal/createAnnotation.container.tsx]: Incorrect spelling for import.
- Per Rush's comment, when the filename is fixed, this should be too.

File: cvat-ui/src/reducers/models-reducer.ts [hunk 5]
```diff
@@ -4,6 +4,7 @@
 
 import { boundariesActions, BoundariesActionTypes } from 'actions/boundaries-actions';
 import { ModelsActionTypes, ModelsActions } from 'actions/models-actions';
+import {CreateAnnotationActionTypes, CretaeAnnotationActions } from 'onepanel/createAnnotationModal/createAnnotation.action';
```
Comment (aleksandrmelnikov) [inline cvat-ui/src/reducers/models-reducer.ts]: Typo: CretaeAnnotationActions

File: cvat-ui/src/reducers/models-reducer.ts [hunk 6]
```diff
@@ -22,7 +23,7 @@ const defaultState: ModelsState = {
 
 export default function (
     state = defaultState,
-    action: ModelsActions | AuthActions | boundariesActions,
+    action: ModelsActions | AuthActions | boundariesActions | CretaeAnnotationActions,
```
Comment (aleksandrmelnikov) [inline cvat-ui/src/reducers/models-reducer.ts]: Typo: CretaeAnnotationActions

File: cvat/apps/onepanelio/views.py [hunk 7]
```diff
@@ -4,28 +4,84 @@
...
+            access_key = yaml.load(file, Loader=yaml.FullLoader)
+            
+        with open("/etc/onepanel/artifactRepositoryS3SecretKey") as file:
+            secret_key = yaml.load(file, Loader=yaml.FullLoader)
+
+        #set env vars
+        os.environ['AWS_ACCESS_KEY_ID'] = access_key
+        os.environ['AWS_SECRET_ACCESS_KEY'] = secret_key
+
+    elif cloud_provider == "gcs":
```
Comment (aleksandrmelnikov) [inline cvat/apps/onepanelio/views.py:53]: Is GCS not supported with this code?

File: cvat/apps/onepanelio/views.py [hunk 8]
```diff
@@ -39,3 +95,248 @@ def get_workflow_templates(request):
...
+
+    configuration = onepanel_authorize()
+    # Enter a context with an instance of the API client
+    with onepanel.core.api.ApiClient(configuration) as api_client:
+        # Create an instance of the API class
+        api_instance = onepanel.core.api.WorkflowTemplateServiceApi(api_client)
+        namespace = os.getenv('ONEPANEL_RESOURCE_NAMESPACE')  # str |
+    try:
+        api_response = api_instance.get_workflow_template2(namespace, uid=workflow_uid, version=str(workflow_version))
+        pprint(api_response)
```
Comment (aleksandrmelnikov) [inline cvat/apps/onepanelio/views.py]: This should be removed once development is done.
The pprint()

File: cvat/apps/onepanelio/views.py [hunk 9]
```diff
@@ -39,3 +95,248 @@ def get_workflow_templates(request):
...
+    configuration = onepanel_authorize()
+
+    # Enter a context with an instance of the API client
+    with onepanel.core.api.ApiClient(configuration) as api_client:
+        # Create an instance of the API class
+        api_instance = onepanel.core.api.ConfigServiceApi(api_client)
+    
+    try:
+        api_response = api_instance.get_config()
+        pprint(api_response)
```
Comment (aleksandrmelnikov) [inline cvat/apps/onepanelio/views.py]: pprint should be removed once development is done.

File: cvat/apps/onepanelio/views.py [hunk 10]
```diff
@@ -39,3 +95,248 @@ def get_workflow_templates(request):
...
+        return JsonResponse(api_response.to_dict()['node_pool']['options'])
+        
+    except ApiException as e:
+        print("Exception when calling ConfigServiceApi->get_config: %s\n" % e)
+
+@api_view(['POST'])
+def get_object_counts(request, pk):
+    # db_task = self.get_object()
```
Comment (aleksandrmelnikov) [inline cvat/apps/onepanelio/views.py:146]: If this is not necessary, please remove commented out code.

File: cvat/apps/onepanelio/views.py [hunk 11]
```diff
@@ -39,3 +95,248 @@ def get_workflow_templates(request):
...
+    data = annotation.get_task_data_custom(pk, request.user)
+    return Response(data)
+
+   
+@api_view(['POST'])
+def get_model_keys(request):
+    form_data = json.loads(request.body.decode('utf-8'))
+    bucket_name = authenticate_cloud_storage()
+    # all_models = [x for x in os.listdir("/home/django/share/models") if os.path.isdir(x)]
```
Comment (aleksandrmelnikov) [inline cvat/apps/onepanelio/views.py:155]: Is this code still needed?

File: cvat/apps/onepanelio/views.py [hunk 12]
```diff
@@ -39,3 +95,248 @@ def get_workflow_templates(request):
...
+    form_data = json.loads(request.body.decode('utf-8'))
+    slogger.glob.info("Form data without preprocessing {} {}".format(form_data, type(form_data)))
+ 
+    form_args = form_data['arguments']
+    time = datetime.now()
+    stamp = time.strftime("%m%d%Y%H%M%S")
+
+    if "cpu" in form_data['machine_type']:
+        tf_image = "tensorflow/tensorflow:1.13.1-py3"
+        machine ="Standard_D4s_v3"
```
Comment (aleksandrmelnikov) [inline cvat/apps/onepanelio/views.py]: I noticed we're hard-coded the machine type here? Are we unable to do this dynamically at this point?

File: cvat/apps/onepanelio/views.py [hunk 13]
```diff
@@ -39,3 +95,248 @@ def get_workflow_templates(request):
...
+            if "base_model" in form_data and "tfod" in form_data['base_model']:
+                ref_model_path = os.getenv('AWS_S3_PREFIX')+'/'+os.getenv('ONEPANEL_RESOURCE_NAMESPACE')+'/'+form_data['base_model']
+            else:
+                ref_model_path = ""
+            params.append(Parameter(name='model-path',value=os.getenv('ONEPANEL_RESOURCE_NAMESPACE')+'/workflow_data/models/'+os.getenv('ONEPANEL_RESOURCE_UID')+'_'+db_task.name+"_tfod_"+stamp+'/'))
+            params.append(Parameter(name='ref-model-path', value=ref_model_path))
+            params.append(Parameter(name='num-classes', value=str(num_classes)))
+            params.append(Parameter(name="ref-model", value=form_data['ref_model']))
+            body = onepanel.core.api.CreateWorkflowExecutionBody(parameters=params,
+            workflow_template_uid = os.getenv('ONEPANEL_OD_TEMPLATE_ID'))
```
Comment (aleksandrmelnikov) [inline cvat/apps/onepanelio/views.py]: Can we retrieve this id from our SDK or not at this time?

File: cvat/apps/onepanelio/views.py [hunk 14]
```diff
@@ -39,3 +95,248 @@ def get_workflow_templates(request):
...
+        # params.append(Parameter(name="source", value="https://github.com/onepanelio/Mask_RNN.git"))
+        params.append(Parameter(name="dataset-path", value=cloud_prefix+dataset_name))
+        params.append(Parameter(name="bucket-name", value=bucket_name))
+        params.append(Parameter(name='task-name', value=db_task.name))
+        # params.append(Parameter(name='num-classes', value=str(num_classes)))
+        params.append(Parameter(name='extras', value=json.dumps(args_and_vals).replace(" ","").replace("{","").replace("}","").replace(":","=")))
+        params.append(Parameter(name="tf-image", value=tf_image))
+        params.append(Parameter(name="sys-node-pool", value=machine))
+        if 'TFRecord' in form_data['dump_format']:
+            if "base_model" in form_data and "tfod" in form_data['base_model']:
```
Comment (aleksandrmelnikov) [inline cvat/apps/onepanelio/views.py]: This if-else seems to be repeated in the else statement below.
I think we can pull this out to reduce code duplication.

File: cvat/apps/tracking/tracker.py [hunk 15]
```diff
@@ -1,101 +1,110 @@
...
 import copy
+from cvat.settings.base import DATA_ROOT
+import os, fnmatch
+from cvat.apps.engine.frame_provider import FrameProvider
+from cvat.apps.engine.models import Task as TaskModel
+from PIL import Image
+import numpy as np
 
 
 def rectanlge_to_cv_bbox(rectangle_points):
```
Comment (aleksandrmelnikov) [inline cvat/apps/tracking/tracker.py]: Typo in function name.

File: cvat/apps/tracking/tracker.py [hunk 16]
```diff
@@ -1,101 +1,110 @@
...
-        self._tracker = trackerTypes_constructor[trackerType]()
+def make_image_list(jid, start_frame, stop_frame):
+	
+	db_task = TaskModel.objects.get(pk=jid)
+	# Get image list
+	image_list = FrameProvider(db_task.data)
+	image_list = image_list.get_frames(image_list.Quality.ORIGINAL)
+	count = 0
+	for img in image_list:
+		# count += 1
```
Comment (aleksandrmelnikov) [inline cvat/apps/tracking/tracker.py]: Un-needed?

File: cvat/apps/tracking/views.py [hunk 17]
```diff
@@ -3,65 +3,80 @@
...
-    tracker = RectangleTracker()
-    new_shapes = tracker.track_rectangles(task, start_shape, stop_frame)
-    slogger.glob.info("new shapes {}".format(new_shapes))
-    new_shapes = [TrackedShapeSerializer(s).data for s in new_shapes]
+@api_view(['POST'])
+@login_required
+@permission_required(perm=['engine.task.access'], raise_exception=True)
+def tracking(request, tid):
+	data = json.loads(request.body.decode('utf-8'))
+	# slogger.glob.info("data {}".format(data))
```
Comment (aleksandrmelnikov) [inline cvat/apps/tracking/views.py]: Not needed?

**Review by aleksandrmelnikov — COMMENTED** [review]
Great work over all.
I'll do some testing of the code.

File: cvat-ui/src/onepanel/createAnnotationModal/createAnnotation.container.tsx (hunk 4 above)
Comment (savan77) [reply cvat-ui/src/onepanel/createAnnotationModal/createAnnotation.container.tsx]: thanks, noted. Not sure if I can tag Naman here, let me try. @TheNaman047

File: cvat/apps/onepanelio/views.py (hunk 14 above)
Comment (savan77) [reply cvat/apps/onepanelio/views.py]: yes, in fact, this whole function will be replaced once we have different payload from the front-end.

File: cvat/apps/tracking/views.py (hunk 17 above)
Comment (savan77) [reply cvat/apps/tracking/views.py]: yes, removed.

File: cvat/apps/tracking/tracker.py (hunk 16 above)
Comment (savan77) [reply cvat/apps/tracking/tracker.py]: yes, removed.

File: cvat/apps/tracking/tracker.py (hunk 15 above)
Comment (savan77) [reply cvat/apps/tracking/tracker.py]: ah,thanks. fixed.

File: cvat/apps/onepanelio/views.py (hunk 13 above)
Comment (savan77) [reply cvat/apps/onepanelio/views.py]: yes, we can.This is the function that needs to be updated based on request payload.

File: cvat/apps/onepanelio/views.py (hunk 8 above)
Comment (savan77) [reply cvat/apps/onepanelio/views.py]: yes, removed.

File: cvat/apps/onepanelio/views.py (hunk 7 above)
Comment (savan77) [reply cvat/apps/onepanelio/views.py:53]: Need to add authentication part.

File: cvat-ui/src/reducers/models-reducer.ts (hunk 5 above)
Comment (savan77) [reply cvat-ui/src/reducers/models-reducer.ts]: cc @TheNaman047

**Pull request merged**